src/sage/topology/moment_angle_complex.py

```diff
@@ -814,3 +819,712 @@ def has_trivial_lowest_deg_massey_product(self):
+                        chmlgy = S.cohomology(deg-i-1, base_ring, generators=True)
+                        for y in chmlgy:
+                            self._gens[deg].append((set(x), deg-i-1, y))
+                        num_of_gens += len(chmlgy)
+                    elif len(S.vertices()) == 0 and deg == 0:
+                        num_of_gens = 1
+
+            indices.extend([(deg, k) for k in range(num_of_gens)])
+            self._graded_indices[deg] = range(num_of_gens)
+
+        cat = Algebras(base_ring).WithBasis().Graded().FiniteDimensional()
+        CombinatorialFreeModule.__init__(self, base_ring, indices, category=cat)
+
+    def basis(self, d=None):
+        """
+        Return (the degree ``d`` homogeneous component of) the basis
+        of ``self``.
+
+        INPUT:
+
+        - ``d`` -- (optional) the degree
+
+        EXAMPLES::
+
+            sage: S2 = simplicial_complexes.Sphere(2)
+            sage: Z = MomentAngleComplex(S2)
+            sage: H = Z.cohomology_ring()
+            sage: H.basis()
+            Finite family {(0, 0): h^{0,0}, (7, 0): h^{7,0}}
+            sage: H.basis(5)
+            Finite family {}
+            sage: H.basis(7)
+            Finite family {(7, 0): h^{7,0}}
+            sage: H.basis()[7, 0]
+            h^{7,0}
+            sage: H.basis(8)
+            Finite family {}
+        """
+        if d is None:
+            return Family(self._indices, self.monomial)
+        else:
+            indices = [(d, i) for i in self._graded_indices.get(d, [])]
+            return Family(indices, self.monomial)
+
+    def degree_on_basis(self, i):
+        """
+        Return the degree of the basis element indexed by ``i``.
+
+        EXAMPLES::
+
+            sage: Z = MomentAngleComplex([[1,2], [3,4,5], [1,5], [2,3]])
+            sage: H = Z.cohomology_ring(GF(5))
+            sage: H.degree_on_basis((4, 0))
+            4
+        """
+        return i[0]
+
+    def _repr_(self):
+        """
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: MomentAngleComplex([[1,2,3], [3,4]]).cohomology_ring()
+            Cohomology module of Moment-angle complex of Simplicial complex
+            with vertex set (1, 2, 3, 4) and facets {(3, 4), (1, 2, 3)}
+            over Rational Field
+        """
+        return "Cohomology module of {} over {}".format(self._complex, self.base_ring())
+
+    def _repr_term(self, i):
+        """
+        Return ``'h^{i[0],i[1]}'``, for the basis element indexed by ``i``.
+
+        EXAMPLES::
+
+            sage: Z = MomentAngleComplex([[1,2], [3,4,5], [1,5], [2,3]])
+            sage: Z = MomentAngleComplex([[1,2], [3,4,5], [1,4], [2,3,5]])
+            sage: H = Z.cohomology_ring(GF(7))
+            sage: H.basis()[3, 2]
+            h^{3,2}
+        """
+        return 'h^{{{},{}}}'.format(i[0], i[1])
+
+    _latex_term = _repr_term
+
+    def one(self):
+        """
```

Comment:
```suggestion

    @cached_method
    def one(self):
        """
```
So we don't have to repeatedly recompute it as it can be called often.

Actually, this is always just a single element $h^{0,0}$ right? If so, then you use should `one_basis()` and return that index `(0, 0)`. This seems to be an assumption of `_to_cycle_on_basis` and `product_on_basis`.